File: app/src/main/java/se/kicksort/metalarchives/model/BandSearchResult.java
package se.kicksort.metalarchives.model;

import com.github.wrdlbrnft.sortedlistadapter.SortedListAdapter;
import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

/**
 * Created by Gustav Sundin on 13/03/17.
 */

public class BandSearchResult implements SortedListAdapter.ViewModel {
    @SerializedName("_id")
    @Expose
    private String id;

    @SerializedName("_score")
    @Expose
    private Double searchScore;

    @SerializedName("_source")
    @Expose
    private SearchResultsInnerData source;

    public String getId() {
        return id;
    }

    public Double getSearchScore() {
        return searchScore;
    }

    public String getBandName() {
        return source.getBandName();
    }

    public String getGenre() {
        return source.getGenre();
    }

    public String getCountry() {
        return source.getCountry();
    }

    public String getStatus() {
        return source.getStatus();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof BandSearchResult)) return false;

        return isSameModelAs(o) && isContentTheSameAs(o);
    }

    @Override
    public int hashCode() {
        int result = source.getBandName() != null ? source.getBandName().hashCode() : 0;
        result = 31 * result + (id != null ? id.hashCode() : 0);
        result = 31 * result + (source.getGenre() != null ? source.getGenre().hashCode() : 0);
        result = 31 * result + (source.getCountry() != null ? source.getCountry().hashCode() : 0);

        return result;
    }

    @Override
    public <T> boolean isSameModelAs(T item) {
        if (item instanceof BandSearchResult) {
            final BandSearchResult other = (BandSearchResult) item;
            return other.id.equals(id);
        }
        return false;
    }

    @Override
    public <T> boolean isContentTheSameAs(T item) {
        if (item instanceof BandSearchResult) {
            final BandSearchResult other = (BandSearchResult) item;
            if (!source.getBandName().equals(other.getBandName())) {
                return false;
            }
            if (!source.getCountry().equals(other.getCountry())) {
                return false;
            }
            return source.getGenre() != null ? source.getGenre().equals(other.getGenre()) : other.getGenre() == null;
        }
        return false;
    }
}

class SearchResultsInnerData {
    @SerializedName("band_name")
    @Expose
    private String name = "";

    @SerializedName("genre")
    @Expose
    private String genre = "";

    @SerializedName("country")
    @Expose
    private String country = "";

    @SerializedName("status")
    @Expose
    private String status = "";

    public String getBandName() {
        return name;
    }

    public String getGenre() {
        return genre;
    }

    public String getCountry() {
        return country;
    }

    public String getStatus() {
        return status;
    }
}

File: app/src/main/java/se/kicksort/metalarchives/model/BandSearchResult.kt
package se.kicksort.metalarchives.model

import com.github.wrdlbrnft.sortedlistadapter.SortedListAdapter
import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName

/**
 * Created by Gustav Sundin on 13/03/17.
 */

class BandSearchResult : SortedListAdapter.ViewModel {
    @SerializedName("_id")
    @Expose
    val id: String? = null

    @SerializedName("_score")
    @Expose
    val searchScore: Double? = null

    @SerializedName("_source")
    @Expose
    private val source: SearchResultsInnerData? = null

    val bandName: String?
        get() = source!!.bandName

    val genre: String?
        get() = source!!.genre

    val country: String?
        get() = source!!.country

    val status: String
        get() = source!!.status

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        return if (o !is BandSearchResult) false else isSameModelAs<Any>(o) && isContentTheSameAs<Any>(o)

    }

    override fun hashCode(): Int {
        var result = if (source!!.bandName != null) source.bandName!!.hashCode() else 0
        result = 31 * result + (id?.hashCode() ?: 0)
        result = 31 * result + if (source.genre != null) source.genre.hashCode() else 0
        result = 31 * result + if (source.country != null) source.country.hashCode() else 0

        return result
    }

    override fun <T> isSameModelAs(item: T): Boolean {
        if (item is BandSearchResult) {
            val other = item as BandSearchResult
            return other.id == id
        }
        return false
    }

    override fun <T> isContentTheSameAs(item: T): Boolean {
        if (item is BandSearchResult) {
            val other = item as BandSearchResult
            if (source!!.bandName != other.bandName) {
                return false
            }
            if (source.country != other.country) {
                return false
            }
            return if (source.genre != null) source.genre == other.genre else other.genre == null
        }
        return false
    }
}

internal class SearchResultsInnerData {
    @SerializedName("band_name")
    @Expose
    val bandName: String? = ""

    @SerializedName("genre")
    @Expose
    val genre: String? = ""

    @SerializedName("country")
    @Expose
    val country: String? = ""

    @SerializedName("status")
    @Expose
    val status = ""
}